### ISSUE TITLE
Staypoint generation can have invalid id assignment for positionfixes (causes error in tripleg generation)

### ISSUE THREAD

**Issue by henrymartin1** [mie-lab/trackintel#229]
When generating staypoints it can happen that a single positionfix in a group of positionfixes that belong to a single staypoint does not get assigned the staypoint id but stays nan. This causes errors in the trip generation as this positionfix leads to the generation of an invalid tripleg with length 1 (The detection mechanism for invalid triplegs depends on the staypoint id). This problem occurs for Geolife users 165, 59, 65, 10, 163, 141. Therefore I am quite sure that this problem has to be related to a recent PR because I have tested the tripleg generation with all geolife users in the past. 

Here is the code to reproduce this problem for geolife user 65: 
```Python
import trackintel as ti
import datetime
import pytz
# the path to a folder that contains geolife user 65
geolife_path = r"xxx"

pfs, mode_labels =ti.io.dataset_reader.read_geolife(geolife_path)
pfs, spts = pfs.as_positionfixes.generate_staypoints(gap_threshold=15, include_last=True, print_progress=True)
start = datetime.datetime(year=2011, month=9, day=3, hour=23, minute=21, second=55, tzinfo=pytz.utc)
end = datetime.datetime(year=2011, month=9, day=3, hour=23, minute=22, second=15, tzinfo=pytz.utc)
pfs_ = pfs[(pfs['tracked_at'] > start) & (pfs['tracked_at'] < end) & (pfs['user_id'] == 65)]

print(pfs_['staypoint_id'])
```
Output: 

|tracked_at | staypoint id |
|---|---|
|234703 2011-09-03 23:21:56+00:00  |        <NA>|
|234704 2011-09-03 23:21:56+00:00  |        <NA>|
|234705 2011-09-03 23:21:56+00:00  |        <NA>|
|234706 2011-09-03 23:21:56+00:00  |        <NA>|
|234707 2011-09-03 23:21:56+00:00  |        <NA>|
|234708 2011-09-03 23:21:57+00:00  |         533|
|234709 2011-09-03 23:21:57+00:00  |         533|
|234710 2011-09-03 23:21:57+00:00  |         533|
|234711 2011-09-03 23:21:57+00:00  |         533|
|234712 2011-09-03 23:21:57+00:00  |        <NA>|
|234713 2011-09-03 23:21:58+00:00  |         533|
|234714 2011-09-03 23:22:14+00:00  |         533|
|234715 2011-09-03 23:22:14+00:00  |         533|
|234716 2011-09-03 23:22:14+00:00  |         533|
|234717 2011-09-03 23:22:14+00:00  |         533|
|234718 2011-09-03 23:22:14+00:00  |         533|

The expected output should be that there is no gap in the staypoint id between positionfixes that belong to the same staypoint.

**Comment by hong2223** [comment]
The problem comes from Line [396](https://github.com/mie-lab/trackintel/blob/4cf96573272bbdc8e475063e3edc1c36b7e44f69/trackintel/preprocessing/positionfixes.py#L396): the default sort method of [`.sort_values` function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html) does not preserve the order of the input when two records are identical. We also need to make sure that the index is ordered. So I proposed to change the line into: 

> df = df.sort_index(kind="mergesort").sort_values(by=["tracked_at"], kind="mergesort")

This solves the problem for user 65, please check the other users.

**Comment by hong2223** [comment]
I had a closer look at this bug. This weird behaviour is generated because：
1.  `.sort_values` function randomly assigns orders for records with the same value.
2. even though the records' `tracked_at` time is the same, their geometry is slightly different (due to GPS noise). 

As our `_generate_staypoints_sliding_user` function process records in a sequential matter after the sorting, it could happen that the first record (after randomly assign order) does not fulfil the `dist_threshold`, but the second does. The result of this is only the second pfs is considered in the staypoint. After restoring the order, the missed `pfs` would appear in the middle of the original df ranked according to the index (just like above). 

So to mitigate this behaviour, we need to make sure the index of pfs after `tracked_at` sorting correspond to the original order of index. That means:  `df = df.sort_values(by=["tracked_at"], kind="mergesort")`
But if the original pfs is randomly ordered and the user sort pfs according to index after running this function, the original problem would appear. Therefore we need  `df = df.sort_index(kind="mergesort").sort_values(by=["tracked_at"], kind="mergesort")`. This does not work for situations where the index is not sortable.

Another solution would be to manually fill in the missing `staypoint id` after the for loop.

**Comment by henrymartin1** [comment]
Ok, I tried your solution and it seems to work, so that's great 👍 
However, there is a catch. I tried for a while but I am not able to create test data for this case. It is really hard to create a small set of positionfixes that leads to the output that I have posted above because we can not simply _cut out_ a portion of the geolife dataset because the staypoint generation depends on the positionfixes that were already parsed. 

The second problem is that  I can also not really test it with the full geolife dataset because #231 keeps throwing erros and it is really hard to filter these without potentially filtering errors that are generated by this bug...

I think the best way would be to add a filter for duplicates + your solution and then run the full geolife again
